### PR TITLE
Feature/acs-29 pipeline-all-amps for repo and share docker image creation

### PR DESCRIPTION
**Pull request by sridharvellingiri**

### CODE REVIEW

**Comment by sridharvellingiri** [comment]
Here is the travis job result for pipeline-all-amps --> https://travis-ci.com/github/Alfresco/acs-packaging/builds/176049476

File: docs/create-pipeline-all-amps-images.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# ACS Pipeline-all-amps Repo & Share Docker Image
```
Comment (alandavis) [inline docs/create-pipeline-all-amps-images.md:1]: Delete this. The same text is in the README

File: tests/environment/docker-compose-pipeline-all-amps.yml [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                -Ddeployment.method=DOCKER_COMPOSE
+                -Dtransform.service.enabled=true
+                -Dtransform.service.url=http://transform-router:8095
+                -Dsfs.url=http://shared-file-store:8099/
+                -DlocalTransform.core-aio.url=http://transform-core-aio:8090/
+                -Dalfresco-pdf-renderer.url=http://transform-core-aio:8090/
+                -Djodconverter.url=http://transform-core-aio:8090/
+                -Dimg.url=http://transform-core-aio:8090/
+                -Dtika.url=http://transform-core-aio:8090/
+                -Dtransform.misc.url=http://transform-core-aio:8090/
```
Comment (alandavis) [inline tests/environment/docker-compose-pipeline-all-amps.yml:51]: These lines should look like lines 42-52 of https://github.com/Alfresco/acs-deployment/blob/master/docker-compose/docker-compose.yml

File: tests/pipeline-all-amps/Readme.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# ACS Pipeline-all-amps Repo & Share Docker Image
```
Comment (alandavis) [inline tests/pipeline-all-amps/Readme.md:1]: Replace with:

# ACS Pipeline-all-amps Repo & Share Docker Images

The [pipeline-all-amps](../pipeline-all-amps) module pushes repo and share images with extra amps applied on top for use by the CI/CD single pipeline process.

## Travis job to test and push docker images

acs-packaging repo [travis](../../.travis.yml) job `"Pipeline All AMPs tests and push docker images"` publishes internally the new docker images for repo and share with travis build number `Example: alfresco/alfresco-pipeline-all-amps-repo:3102:7.0.0`. The images has been tested using [docker-compose](../tests/environment/docker-compose-pipeline-all-amps.yml) deployment.

Repo and share images are pushed to [Quay docker repo](https://quay.io/repository/alfresco/alfresco-pipeline-all-amps-repo?tab=info) and [Quay docker share](https://quay.io/repository/alfresco/alfresco-pipeline-all-amps-share?tab=info)

File: tests/pipeline-all-amps/repo/.maven-dockerignore [hunk 4]
```diff
@@ -0,0 +1 @@
+target/docker/**
```
Comment (alandavis) [inline tests/pipeline-all-amps/repo/.maven-dockerignore:1]: This should be in the top level file.

File: tests/pipeline-all-amps/repo/pom.xml [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+  <properties>
+    <image.name>alfresco/alfresco-pipeline-all-amps-repo</image.name>
+
+    <alfresco.alfresco-rm-enterprise-rest-api-explorer.version>3.0.2</alfresco.alfresco-rm-enterprise-rest-api-explorer.version>
+  </properties>
+
+<dependencyManagement>
+  <dependencies>
+    <dependency>
```
Comment (alandavis) [inline tests/pipeline-all-amps/repo/pom.xml:20]: If any of these exist in the top level pom, remove them from here.

File: tests/pipeline-all-amps/repo/pom.xml [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+  </dependencies>
+</dependencyManagement>
+
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-dependency-plugin</artifactId>
+        <version>3.0.2</version>
```
Comment (alandavis) [inline tests/pipeline-all-amps/repo/pom.xml:101]: use 3.1.2

File: tests/pipeline-all-amps/repo/pom.xml [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        <property>
+          <name>!skipS3Connector</name>
+        </property>
+      </activation>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-dependency-plugin</artifactId>
+            <version>3.0.2</version>
```
Comment (alandavis) [inline tests/pipeline-all-amps/repo/pom.xml:168]: use 3.1.2

File: tests/pipeline-all-amps/share/.maven-dockerignore [hunk 8]
```diff
@@ -0,0 +1 @@
+target/docker/**
```
Comment (alandavis) [inline tests/pipeline-all-amps/share/.maven-dockerignore:1]: Should be in top level file

File: tests/pipeline-all-amps/share/pom.xml [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    </dependencies>
+  </dependencyManagement>
+
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-dependency-plugin</artifactId>
+        <version>3.0.2</version>
```
Comment (alandavis) [inline tests/pipeline-all-amps/share/pom.xml:59]: use 3.1.2

File: .travis.yml [hunk 10]
```diff
@@ -182,6 +182,21 @@ jobs:
...
         - travis_wait 20 mvn install -f tests/tas-all-amps/pom.xml -Pall-tas-tests -Denvironment=default -DrunBugs=false
+    - name: "Pipeline All AMPs tests and push docker images"
+      jdk: openjdk11
+      if: fork = false AND (branch = master OR branch =~ /support\/.*/) AND type != pull_request
+      install:
+        # Build additional Docker image for repo and share for CI pipeline
+        - export TRAVIS_BUILD_NUMBER=${TRAVIS_BUILD_NUMBER}
+        - travis_retry travis_wait 20 mvn install -q -f tests/pipeline-all-amps/pom.xml -DskipTests
+        - docker login quay.io -u ${QUAY_USERNAME} -p ${QUAY_PASSWORD}
+        - travis_retry travis_wait 20 mvn clean install -f ./tests/pipeline-all-amps/pom.xml deploy -Prelease
```
[the remark this answers is not in the document]
Comment (sridharvellingiri) [reply .travis.yml:193]: > Looking a lot better.
> 
> * Have you checked that the correct images are pushed to https://quay.io/repository/?
>   I don't see them, or rather I do do see the wrong one for share from 4 days ago.
>   It also appears never to be deleted. It should be deleted after a few days as it will be
>   of little value after that.
> * Have you checked that the image is NOT pushed to https://hub.docker.com/u/alfresco/ ?

The new build images are pushed to https://quay.io/repository/alfresco/alfresco-pipeline-all-amps-repo?tab=tags for repo and https://quay.io/repository/alfresco/alfresco-pipeline-all-amps-share?tab=info for share.

The images not pushed to Docker hub.

**Comment by sridharvellingiri** [comment]
> Looking a lot better.
> 
> * Have you checked that the correct images are pushed to https://quay.io/repository/?
>   I don't see them, or rather I do do see the wrong one for share from 4 days ago.
>   It also appears never to be deleted. It should be deleted after a few days as it will be
>   of little value after that.
> * Have you checked that the image is NOT pushed to https://hub.docker.com/u/alfresco/ ?

File: tests/environment/docker-compose-pipeline-all-amps.yml [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+                -Xmx500m
+                -Dalfresco.host=localhost
+                -Dalfresco.port=8080
+                -Dalfresco.context=alfresco
+                -Dalfresco.protocol=http
+                "
+            TRAVIS_BUILD_NUMBER: "${TRAVIS_BUILD_NUMBER}"
+
+    postgres:
+        image: postgres:11.4
```
Comment (gavincornwell) [inline tests/environment/docker-compose-pipeline-all-amps.yml:122]: I believe we're moving to Postgres 11.7 for the latest releases so please update this reference. As Alan says this file should be as close as possible to https://github.com/Alfresco/acs-deployment/blob/master/docker-compose/docker-compose.yml

File: tests/environment/docker-compose-pipeline-all-amps.yml [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+        image: alfresco/alfresco-activemq:5.15.8
+        mem_limit: 1g
+        ports:
+            - 8161:8161 # Web Console
+            - 5672:5672 # AMQP
+            - 61616:61616 # OpenWire
+            - 61613:61613 # STOMP
+
+    digital-workspace:
+        image: quay.io/alfresco/alfresco-digital-workspace:1.5.0
```
Comment (gavincornwell) [inline tests/environment/docker-compose-pipeline-all-amps.yml:160]: Should be 1.6.0

File: tests/environment/docker-compose-pipeline-all-amps.yml [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+            - digital-workspace
+        ports:
+            - 8080:8080
+        links:
+            - digital-workspace
+            - alfresco
+            - share
+
+    sync-service:
+        image: quay.io/alfresco/service-sync:3.3.2
```
Comment (gavincornwell) [inline tests/environment/docker-compose-pipeline-all-amps.yml:179]: Should be 3.3.3

File: .travis.yml [hunk 14]
```diff
@@ -182,6 +182,21 @@ jobs:
...
+      script:
+        - export TRAVIS_BUILD_NUMBER=${TRAVIS_BUILD_NUMBER}
+        - export PROJECT_VERSION=`mvn help:evaluate -f ./tests/pipeline-all-amps/pom.xml -Dexpression=project.version -q -DforceStdout`
```
[the remark this answers is not in the document]
Comment (sridharvellingiri) [reply .travis.yml:196]: Updated with single mvn command. The extra mvn build command has been removed.
mvn help:evaluate command used to export PROJECT_VERSION to use it in the docker-compose. Im not sure these commands can be combined

File: README.md [hunk 15]
```diff
@@ -52,3 +52,5 @@ In this folder the distribution zip is built. It contains all the war files, lib
 ## How to
 
 * [Create a custom Docker image](docs/create-custom-image.md)
+
```
[the remark this answers is not in the document]
Comment (sridharvellingiri) [reply README.md]: Removed Pipeline-all-amps line from README.md

File: tests/environment/docker-compose-pipeline-all-amps.yml (hunk 2 above)
Comment (sridharvellingiri) [reply tests/environment/docker-compose-pipeline-all-amps.yml:51]: Both compose have same values

File: tests/pipeline-all-amps/Readme.md (hunk 3 above)
Comment (sridharvellingiri) [reply tests/pipeline-all-amps/Readme.md:1]: Done

File: tests/pipeline-all-amps/repo/.maven-dockerignore (hunk 4 above)
Comment (sridharvellingiri) [reply tests/pipeline-all-amps/repo/.maven-dockerignore]: moved this to pipeline-all-amps

File: tests/pipeline-all-amps/repo/pom.xml (hunk 5 above)
Comment (sridharvellingiri) [reply tests/pipeline-all-amps/repo/pom.xml]: Done

File: tests/pipeline-all-amps/share/.maven-dockerignore (hunk 8 above)
Comment (sridharvellingiri) [reply tests/pipeline-all-amps/share/.maven-dockerignore]: done

File: docs/create-pipeline-all-amps-images.md (hunk 1 above)
Comment (sridharvellingiri) [reply docs/create-pipeline-all-amps-images.md]: Done

File: tests/pipeline-all-amps/repo/pom.xml (hunk 6 above)
Comment (sridharvellingiri) [reply tests/pipeline-all-amps/repo/pom.xml]: Done

File: tests/pipeline-all-amps/share/pom.xml (hunk 9 above)
Comment (sridharvellingiri) [reply tests/pipeline-all-amps/share/pom.xml]: Done

File: tests/pipeline-all-amps/repo/pom.xml (hunk 7 above)
Comment (sridharvellingiri) [reply tests/pipeline-all-amps/repo/pom.xml]: Done

File: tests/environment/docker-compose-pipeline-all-amps.yml (hunk 11 above)
Comment (sridharvellingiri) [reply tests/environment/docker-compose-pipeline-all-amps.yml]: Done

File: tests/environment/docker-compose-pipeline-all-amps.yml (hunk 13 above)
Comment (sridharvellingiri) [reply tests/environment/docker-compose-pipeline-all-amps.yml]: Done

File: tests/environment/docker-compose-pipeline-all-amps.yml (hunk 12 above)
Comment (sridharvellingiri) [reply tests/environment/docker-compose-pipeline-all-amps.yml]: Done

**Comment by sridharvellingiri** [comment]
The latest travis job for this change -- > https://travis-ci.com/github/Alfresco/acs-packaging/builds/176318324

File: tests/pipeline-all-amps/repo/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,34 @@
+FROM alfresco/alfresco-content-repository:${image.tag}
```
Comment (alandavis) [inline tests/pipeline-all-amps/repo/Dockerfile:1]: Can we just replace the ${image.tag} with latest
This is what is done in the alfresco-with-jolokia and custom-workflow Docker files.
${image.tag} is normally used for the target rather than the source version.

File: tests/tas-all-amps/Dockerfile [hunk 17]
```diff
@@ -1,6 +1,6 @@
 # Fetch image based on Tomcat 8.5.43 and Java 11
 # More infos about this image: https://github.com/Alfresco/alfresco-docker-base-tomcat
-FROM alfresco/alfresco-content-repository:latest
+FROM alfresco/alfresco-content-repository:${image.tag}
```
Comment (alandavis) [inline tests/tas-all-amps/Dockerfile:3]: Please revert this change back to using :latest

**Comment by sridharvellingiri** [comment]
> Approved subject to
> 
> 1. switching to using :latest in two Dockerfiles
> 2. Using the text I provided in the MD file, unless there is a good reason for not doing so.
> 3. That it can be demonstrated that the Image can be used on a single pipeline feature branch.

Done
